src/libjasper/jpc/jpc_t2dec.c

```diff
@@ -189,15 +189,13 @@ static int jpc_dec_decodepkt(jpc_dec_t *dec, jas_stream_t *pkthdrstream, jas_str
 
 	/* Decode the SOP marker segment if present. */
 	if (cp->csty & JPC_COD_SOP) {
-		if (jpc_dec_lookahead(in) == JPC_MS_SOP) {
+		if (jpc_dec_lookahead(in) != JPC_MS_SOP) {
+			jas_eprintf("warning: missing SOP marker segment\n");
+			dec->curtileendoff -= 6; // 6 = SOP(2) + Lsop(2) + Nsop(2)
```

Comment:
Can this underflow?

Reply:
if "underflow" means `dec->curtileendoff` can be negative by this decrement, the answer is yes with a broken input.
Maybe the following checking of "underflow" is necessary.
```
if (dec->curtileendoff < 0) {
   return -1;
}
```

Reply:
Because of all those fuzzers people will create a broken input :) This is why we had to fix all those CVEs. So we should add such checks for new code right from the start.

Reply:
Thanks. I'll modify 1f2987c to include the underflow checking.

Reply:
The commit 1f2987c was updated with the underflow check as you can see as 0fb7332.

Reply:
I got an error on build check but I think the error is out of the scope of this PR. The error was occured at linking for `imgcmp` and it said
```
undefined reference to `__builtin_mul_overflow'
```

Reply:
My PR https://github.com/jasper-software/jasper/pull/229 fixes this linker problem.
What I still don't understand is why you remove 6 from `curtileendoff`. Your explanation sounds like Jasper had assumed it was there and so it had considered this number of bytes in the expected end offset, and now must remove it again. But I don't see such a think in the code:
https://github.com/jasper-software/jasper/blob/9580dcb0b40a8ebe1bd45976235f2f518d0adc0b/src/libjasper/jpc/jpc_dec.c#L527-L528
I don't quite get this formula yet - I don't know what the `- 4` means. But I don't see any `6` here which could account for the SOP segment. And from my understanding of the standard, it would be wrong to assume any SOP length here, because for `SOT.Psot`, it should be irrelevant whether a SOP exists or not - if it exists, its size will be included in `Psot`, and not if there's no such SOP. Why should the expected end offset be changed here?

Reply:
> My PR #229 fixes this linker problem.

Thanks for letting me know.

> What I still don't understand is why you remove 6 from `curtileendoff`. Your explanation sounds like Jasper had assumed it was there and so it had considered this number of bytes in the expected end offset, and now must remove it again. But I don't see such a think in the code:
> https://github.com/jasper-software/jasper/blob/9580dcb0b40a8ebe1bd45976235f2f518d0adc0b/src/libjasper/jpc/jpc_dec.c#L527-L528
> 
> 
> I don't quite get this formula yet - I don't know what the `- 4` means. But I don't see any `6` here which could account for the SOP segment. And from my understanding of the standard, it would be wrong to assume any SOP length here, because for `SOT.Psot`, it should be irrelevant whether a SOP exists or not - if it exists, its size will be included in `Psot`, and not if there's no such SOP. Why should the expected end offset be changed here?

I found subtracting 6 from `dec->curtileendoff` was wrong because I tested the code with a broken codestream which has incorrect `Psot` value. However, I believe the currect treatment of SOP in JasPer is not correct. So, I'll remove the commit for the SOP checking from this PR and I will consider how to correct this problem in the future commit.  

FYI: the `-4` in the formula you put as a link can be considered as `-2` for marker segment and `-2` for a length parameter for a marker segment. Every marker (except for SOC, SOD, and EOC) defined in the spec has `Lxxx(xxx is the name of the marker)` as 2 bytes parameter.

Reply:
Thanks, I think the EPH commit looks simple enough and correct, I'll merge it (and then wait for your SOP fix, which appears to be more complicated).